sample_usage/stats_analysis.py

```diff
@@ -0,0 +1,160 @@
+from collections import OrderedDict
+
+def indiv_calc(error_msgs, style_msgs):
+    """
+    Analyses the given lists of error and style Message objects error_msgs and
+    style_msgs for an individual.
+
+    @param list error_msgs: Message objects for all of a student's code errors
+    @param list style_msgs: Message objects for all of a student's style issues
+    @rtype: list[tuple]
+    """
+
+    # {msg.symbol + "(" + msg.object + ")": count}
+    all_msgs = error_msgs + style_msgs
+
+    all_num = _calc_helper(all_msgs)
+    error_num = _calc_helper(error_msgs)
+    style_num = _calc_helper(style_msgs)
+
+    stats = [('Most Frequent Messages In Numbers', all_num[0]),
+             ('Most Frequent Messages In Percentages', all_num[1]),
+             ('Most Frequent Errors In Numbers', error_num[0]),
+             ('Most Frequent Errors in Percentages', error_num[1]),
+             ('Most Frequent Styles In Numbers', style_num[0]),
+             ('Most Frequent Styles in Percentages', style_num[1])]
+
+    return stats
+
+
+def summary(all_msgs):
+    """
+    Provides a summary of each individual's errors from all_msgs and provide an
+    overall summary of the course's performance (if applicable).
+    Called by pyta_statistics.
+
+    @param dict[str -> tuple(list[Message], list[Message])] all_msgs:
+        the tuple of code and error messages for each student's files
+    @rtype: list[list[tuple]]]
+    """
+    num_stu = len(all_msgs)
+
+    # If directory was for student, not course, return empty summary stats list.
+    if num_stu == 1:
+        student, stats = all_msgs.popitem()
+        return [[(student, indiv_calc(*stats))], []]
+
+    sum_calc = [[], []]     # two * OrderedDict()
+    code_errors = []
+    style_errors = []
+    stu_errors = []
+
+    for student in all_msgs:
+        # in the form [('std1', [<stats>]), ('std2', [<stats>])]
+        errors = all_msgs[student][0]
+        styles = all_msgs[student][1]
+        calc = indiv_calc(errors, styles)
+        sum_calc[0].append((student, calc))
+        # for OrderedDict: indiv_dict[student] = calc
+        stu_errors.append((len(errors) + len(styles), student))
+
+        # To find Most Frequent Errors (aggregate)
+        code_errors.append(errors)
+        style_errors.append(styles)
+
+    error_num = frequent_messages(_message_counter(code_errors))
+    style_num = frequent_messages(_message_counter(style_errors))
+
+    # Calculating the Five Number Summary for all errors (per student)
+    stu_errors.sort(reverse=True)
+
+    if len(stu_errors) == 1:
```

Comment:
For this part, import `statistics` and use [statistics.median](https://docs.python.org/3/library/statistics.html#statistics.median) rather than computing it yourself.